homeassistant/components/fritzbox/__init__.py

```diff
@@ -51,40 +59,51 @@
 )
 
 
-def setup(hass, config):
-    """Set up the fritzbox component."""
+async def async_setup(hass, config):
+    """Set up the AVM Fritz!Box integration."""
+    if DOMAIN in config:
+        for entry_config in config[DOMAIN][CONF_DEVICES]:
+            hass.async_create_task(
+                hass.config_entries.flow.async_init(
+                    DOMAIN, context={"source": "import"}, data=entry_config
+                )
+            )
+
+    return True
+
 
-    fritz_list = []
+async def async_setup_entry(hass, entry):
+    """Set up the AVM Fritz!Box platforms."""
+    fritz = Fritzhome(
+        host=entry.data[CONF_HOST],
+        user=entry.data[CONF_USERNAME],
+        password=entry.data[CONF_PASSWORD],
+    )
+    await hass.async_add_executor_job(fritz.login)
 
-    configured_devices = config[DOMAIN].get(CONF_DEVICES)
-    for device in configured_devices:
-        host = device.get(CONF_HOST)
-        username = device.get(CONF_USERNAME)
-        password = device.get(CONF_PASSWORD)
-        fritzbox = Fritzhome(host=host, user=username, password=password)
-        try:
-            fritzbox.login()
-            _LOGGER.info("Connected to device %s", device)
-        except LoginError:
-            _LOGGER.warning("Login to Fritz!Box %s as %s failed", host, username)
-            continue
+    hass.data.setdefault(DOMAIN, {CONF_CONNECTIONS: {}, CONF_DEVICES: set()})
+    hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id] = fritz
 
-        fritz_list.append(fritzbox)
+    for domain in SUPPORTED_DOMAINS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, domain)
+        )
 
-    if not fritz_list:
-        _LOGGER.info("No fritzboxes configured")
-        return False
+    def logout_fritzbox(event):
+        """Close connections to this fritzbox."""
+        fritz.logout()
 
-    hass.data[DOMAIN] = fritz_list
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzbox)
+
+    return True
 
-    def logout_fritzboxes(event):
-        """Close all connections to the fritzboxes."""
-        for fritz in fritz_list:
-            fritz.logout()
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, logout_fritzboxes)
+async def async_unload_entry(hass, entry):
+    """Unloading the AVM Fritz!Box platforms."""
+    fritz = hass.data[DOMAIN][CONF_CONNECTIONS][entry.entry_id]
+    await hass.async_add_executor_job(fritz.logout)
 
     for domain in SUPPORTED_DOMAINS:
-        discovery.load_platform(hass, domain, DOMAIN, {}, config)
+        await hass.config_entries.async_forward_entry_unload(entry, domain)
```

Comment:
Please use `asyncio.gather` to await all tasks concurrently.

https://github.com/home-assistant/core/blob/7f06a939acefdcdf56d33cbb30ed7688edb31513/script/scaffold/templates/config_flow/integration/__init__.py#L36-L49